### PR TITLE
feat: set GKE disk type labels on Karpenter nodes

### LINKED ISSUE #402: GCENodeClass not setting disk-type labels on nodes via kube-env
**What happened**:
GCENodeClass not setting disk-type labels on nodes via kube-env.
Using a dynamic storageclass with a PVC uses label matching on `disk-type.gke.io/<disk_type>=true` to find nodes that can mount the PVC. Without GCENodeClass setting disk-type labels, that PVC cannot match a karpenter created node.

**What you expected to happen**:
Expect mechanism to set disk-types on resultant node.
Either extended metadata argument support or forwarding existing disk categories from disk specs.
```yaml
 metadata:
    kube-labels: "disk-type.gke.io/hyperdisk-balanced=true"
```
or copy category from these definitions
```yaml
 disks:
    - category: hyperdisk-balanced
      boot: true
```

**How to reproduce it (as minimally and precisely as possible)**:
```yaml
  ---
  apiVersion: storage.k8s.io/v1
  kind: StorageClass
  metadata:
    name: dynamic-rwo
    annotations:
      storageclass.kubernetes.io/is-default-class: "true"
  provisioner: pd.csi.storage.gke.io
  volumeBindingMode: WaitForFirstConsumer
  reclaimPolicy: Delete
  allowVolumeExpansion: true
  parameters:
    type: dynamic
    use-allowed-disk-topology: "true"
    hyperdisk-type: hyperdisk-balanced
    pd-type: pd-balanced
  ---
  apiVersion: karpenter.k8s.gcp/v1alpha1
  kind: GCENodeClass
  metadata:
    name: default
  spec:
    imageSelectorTerms:
      - alias: ContainerOptimizedOS@latest
    disks:
      - category: hyperdisk-balanced
        boot: true
        sizeGiB: 100
  ---
  apiVersion: karpenter.sh/v1
  kind: NodePool
  metadata:
    name: default
  spec:
    template:
      spec:
        nodeClassRef:
          group: karpenter.k8s.gcp
          kind: GCENodeClass
          name: default
        requirements:
          - key: karpenter.sh/capacity-type
            operator: In
            values: ["on-demand"]
          - key: karpenter.k8s.gcp/instance-family
            operator: In
            values: ["n4", "n4d"]
  ---
  apiVersion: v1
  kind: PersistentVolumeClaim
  metadata:
    name: dynamic-topology
  spec:
    accessModes:
      - ReadWriteOnce
    resources:
      requests:
        storage: 4Gi
    volumeMode: Filesystem
    storageClassName: dynamic-rwo
  ---
  apiVersion: v1
  kind: Pod
  metadata:
    name: dynamic-topology
  spec:
    containers:
      - name: pause
        image: registry.k8s.io/pause:3.9
        volumeMounts:
          - name: data
            mountPath: /data
    volumes:
      - name: data
        persistentVolumeClaim:
          claimName: dynamic-topology
```

**Anything else we need to know?**:
Fantastic product, almost production ready!

**Environment**:

- Karpenter-provider-gcp: 0.3.3
- GKE version: 1.35.3-gke.1389002
- Others:

### PR DESCRIPTION
**Pull request by dm3ch**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Adds GKE PD CSI disk-type compatibility labels for Karpenter-provisioned nodes.

This PR:
- sets `disk-type.gke.io/*=true` in `kube-labels` and kube-env `--node-labels`,
- exposes the same labels in `InstanceType.Requirements` for pre-launch scheduling,
- registers the labels as well-known Karpenter labels,
- updates PDCSI compatibility data through `make update`, sourced from `hack/pdcsi-data`,
- adds positive/negative storage e2e coverage for disk-type label scheduling,
- documents disk-type scheduling and the PD CSI `use-allowed-disk-topology` opt-in.

#### Related proposal (if applicable):

N/A

#### Which issue(s) this PR fixes:

Fixes #402

#### Docs and examples

- [x] `docs/` and `examples/` updated (or this PR does not affect user-facing behaviour, configuration, or APIs)
- [x] `MIGRATION.md` updated (or this PR does not require any migration steps)

#### Special notes for your reviewer:

- PDCSI compatibility data is pinned by `hack/pdcsi-data/go.mod` and copied by `make update`.
- Non-GPU e2e suites passed on commit `5c2f1510`; later commits are docs/update-flow changes validated with focused provider tests and `make docs-lint`.
- This PR was prepared with AI assistance (Claude Code). All changes have been reviewed and verified by the author.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds GKE PD CSI disk-type compatibility labels (`disk-type.gke.io/*=true`) to Karpenter-provisioned GKE nodes. Compatibility data is sourced from the upstream PDCSI node-labeler ConfigMap, vendored via a dedicated `hack/pdcsi-data` Go module and refreshed by `make update`.

- **Label injection**: `SetDiskTypeLabels` idempotently strips inherited disk-type labels and writes the correct set for the instance type's machine family into both `kube-labels` and `kube-env --node-labels=`.
- **Scheduling requirements**: `computeRequirements` marks every known disk-type label as `DoesNotExist` by default, then upgrades supported labels for the matched machine family to `In: [\"true\"]`, allowing pre-launch pod scheduling against disk-type nodeSelectors.
- **Well-known registration**: `disktype.AllLabels()` is inserted into `karpv1.WellKnownLabels` at startup, enabling Karpenter's DaemonSet overhead simulation to count pods using these labels as selectors.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; the change adds a new metadata-patching step that is well-tested, correctly ordered after all other patches, and gracefully handles unknown machine families by stripping any inherited disk-type labels rather than propagating potentially wrong ones.

The compatibility data load panics at startup if the embedded YAML is malformed — correct fail-fast behaviour — and the YAML is vendored and tested. The label-injection path is covered by unit tests across all three affected layers (disktype, metadata, instance), including edge cases for multiline KUBELET_ARGS, nil labels, and idempotent re-application. No runtime logic errors were found.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| pkg/providers/disktype/compatibility.go | New package: loads the upstream PDCSI node-labeler ConfigMap via embed.FS and exposes AllLabels(), LabelsForFamily(), and LabelsForInstanceType(). Logic is clean; LabelsForFamily returns a defensive copy, preventing callers from mutating the cached slice. |
| pkg/providers/metadata/utils.go | Adds SetDiskTypeLabels, replaceDiskTypeLabelsInKubeEnv, and replaceDiskTypeLabelsInNodeLabels; all three regexes are now package-level. Logic correctly handles the empty-value edge case (--node-labels= with all disk-type labels stripped) via the \S* regex and strips/re-adds labels idempotently. |
| pkg/providers/instance/instance.go | SetDiskTypeLabels is called last in setupInstanceMetadata (after ApplyCustomMetadata and GPU label injection), ensuring disk-type labels always reflect the actual instance type family regardless of user-supplied metadata overrides. |
| pkg/providers/instancetype/types.go | computeRequirements seeds all known disk-type labels as DoesNotExist, then upgrades the applicable family's labels to In:["true"] via the established Get().Insert() pattern also used for GPU labels. |
| pkg/apis/v1alpha1/labels.go | Registers all disk-type labels from AllLabels() as well-known Karpenter labels at init time, enabling DaemonSet overhead simulation for pods with disk-type nodeSelectors. |
| pkg/providers/disktype/compatibility_test.go | Good unit coverage: Family extraction, LabelsForFamily/LabelsForInstanceType happy-paths, unknown-family nil return, AllLabels sorted invariant, and a defensive-copy test. |
| pkg/providers/metadata/utils_test.go | New tests cover: replacement of inherited labels, multiline KUBELET_ARGS continuation lines, and the nil-labels (strip-all) case; all pass with correct assertions. |
| pkg/providers/instance/instance_test.go | TestBuildInstance_DiskTypeLabels exercises the full buildInstance path for e2-standard-4, confirming correct e2 labels are injected and non-e2 labels inherited from the template are stripped. |

</details>

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant IC as instance.Create()
    participant SIM as setupInstanceMetadata()
    participant DT as disktype pkg
    participant MU as metadata/utils.go

    IC->>SIM: buildInstance(instanceType, nodeClass, ...)
    SIM->>SIM: RemoveGKEBuiltinLabels
    SIM->>SIM: SetMaxPodsPerNode
    SIM->>SIM: PatchUnregisteredTaints
    SIM->>SIM: patchKubeEnv (arch/family/OS)
    SIM->>SIM: AppendNodeClaimLabel / AppendRegisteredLabel
    SIM->>SIM: ApplyCustomMetadata (user overrides)
    SIM->>SIM: setupGPUMetadata (GPU labels)
    SIM->>DT: LabelsForInstanceType(instanceType.Name)
    DT-->>SIM: "map[disk-type.gke.io/*]=true, ok"
    SIM->>MU: SetDiskTypeLabels(metadata, diskLabels)
    MU->>MU: replaceDiskTypeLabelsInNodeLabels(kube-labels)
    MU->>MU: replaceDiskTypeLabelsInKubeEnv(kube-env)
    MU-->>SIM: nil (success)
    SIM-->>IC: instance ready
```
</details>

<sub>Reviews (4): Last reviewed commit: ["chore: reuse kube-env node labels regex"](https://github.com/cloudpilot-ai/karpenter-provider-gcp/commit/a0c03978c68e267758ccb38aaefd7f0230e40bf8) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=35469296)</sub>

### CODE REVIEW

File: pkg/providers/instance/instance.go [hunk 1]
```diff
@@ -860,12 +861,19 @@ func (p *DefaultProvider) setupInstanceMetadata(ctx context.Context, instanceMet
...
+	// GPU labels override both the base template value and any spec.metadata entry.
 	if err := setupGPUMetadata(instanceMetadata, nodeClass, instanceType, isGPUInstance); err != nil {
 		return fmt.Errorf("failed to inject GPU metadata labels: %w", err)
 	}
 
+	diskLabels, ok := disktype.LabelsForInstanceType(instanceType.Name)
+	if !ok {
+		log.FromContext(ctx).V(1).Info("no PD disk compatibility labels found for instance type", "instanceType", instanceType.Name)
+	}
+	if err := metadata.SetDiskTypeLabels(instanceMetadata, diskLabels); err != nil {
```
Comment (jwcesign) [inline pkg/providers/instance/instance.go:875]: This sets the disk-type labels on the launched VM metadata, but Karpenter also needs to know these labels during scheduling. With `use-allowed-disk-topology: "true"`, PD CSI can put `disk-type.gke.io/*` into the PV topology/node affinity; when that pod is rescheduled, Karpenter converts those PV constraints into volume requirements before this VM exists. Since the provider does not expose known values for these custom labels on the `InstanceType.Requirements` (and they are not registered as well-known labels), provisioning can fail as an unknown custom label or fail to filter out unsupported machine families. Can we add the disk-type requirements to `computeRequirements` as well, e.g. supported labels `In ["true"]` and unsupported labels `DoesNotExist`, and register the disk-type labels as well-known?

Comment (dm3ch) [reply pkg/providers/instance/instance.go:875]: Should be fixed in https://github.com/cloudpilot-ai/karpenter-provider-gcp/pull/404/changes/2d138ffab99832cd0ffe178a6db17737ac7eced3

**Comment by dm3ch** [comment]
### E2E Test Results

**Cluster:** `europe-west4-a` | **Commit:** `5c2f1510` | **Total duration:** `~38m`  
**Run mode:** non-GPU suites in parallel with quota cap; provisioning rerun separately with `--procs=1` after the wrapper was interrupted before recording its result.

| Suite | Spec | Result | Duration | Notes |
|---|---|---|---|---|
| Channel image selection | 2 specs | ✅ pass | 5m 8s | |
| Confidential | 2 specs | ✅ pass | 4m 54s | |
| Consolidation | 1 spec | ✅ pass | 3m 20s | |
| Drift | 2 specs | ✅ pass | 6m 53s | |
| Expiration | 1 spec | ✅ pass | 5m 29s | |
| GC | 1 spec | ✅ pass | 5m 56s | |
| Kubelet config | 5 specs | ✅ pass | 11m 56s | |
| Networking | 1 spec | ✅ pass | 2m 3s | |
| Provisioning | 11 specs | ✅ pass | 25m 42s | Rerun after interrupted wrapper; passed with `--procs=1`. |
| Repair | 1 spec | ✅ pass | 2m 51s | |
| Storage | 3 specs | ✅ pass | 6m 31s | Includes disk-type label scheduling tests. |

**Overall: ✅ all non-GPU suites pass** (31 specs, 0 failures). GPU suite intentionally not run.

Controller log scan found no panics. It did contain expected cleanup races (`NodePool ... not found` after test cleanup) and the expected unschedulable event from the negative disk-label storage test (`disk-type.gke.io/hyperdisk-throughput In [true]`).

**Comment by promptless-for-oss** [comment]
Promptless prepared a documentation update related to this change.

Triggered by [PR #404](https://github.com/cloudpilot-ai/karpenter-provider-gcp/pull/404)

Added a new "Disk type scheduling" section to the advanced examples documentation, explaining how Karpenter applies `disk-type.gke.io/*` labels to nodes based on machine family and how to use NodePool requirements to constrain instance types by disk support.

Review: [docs: add disk type scheduling section](https://github.com/cloudpilot-ai/karpenter-provider-gcp/pull/423)

**Comment by dm3ch** [comment]
@promptless-for-oss your doc changes have been merged into this PR

**Comment by promptless-for-oss** [comment]
Thanks for letting me know! Happy to see the documentation updates incorporated directly into the feature PR.

File: pkg/providers/metadata/utils.go [hunk 2]
```diff
@@ -39,6 +40,8 @@ var (
 	gkeProvisioningRegex    = regexp.MustCompile(`gke-provisioning=\w+`)
 	kubeEnvArchRegex        = regexp.MustCompile(`\barch=(amd64|arm64)\b`)
 	kubeEnvFamilyRegex      = regexp.MustCompile(`cloud\.google\.com/machine-family=[^,;\s]+`)
+	diskTypeLabelRegex      = regexp.MustCompile(`^disk-type\.gke\.io/`)
+	diskTypeLabelEntryRegex = regexp.MustCompile(`,?disk-type\.gke\.io/[^=,\s]+=[^,\s]*`)
```
[the remark this answers is not in the document]
Comment (dm3ch) [reply pkg/providers/metadata/utils.go]: Fixed in b271f9a0.

File: pkg/providers/metadata/utils.go [hunk 3]
```diff
@@ -427,6 +430,73 @@ func appendNodeLabelToKubeEnv(metadata *compute.Metadata, labelKey, labelValue s
...
+	return nil
+}
+
+func replaceDiskTypeLabelsInKubeEnv(kubeEnv string, labels map[string]string) (string, error) {
+	if !strings.Contains(kubeEnv, "--node-labels=") {
+		return "", fmt.Errorf("--node-labels flag not found in kube-env KUBELET_ARGS")
+	}
+	stripped := diskTypeLabelEntryRegex.ReplaceAllString(kubeEnv, "")
+	stripped = strings.ReplaceAll(stripped, "--node-labels=,", "--node-labels=")
+	return regexp.MustCompile(`--node-labels=\S*`).ReplaceAllStringFunc(stripped, func(match string) string {
```
[the remark this answers is not in the document]
Comment (dm3ch) [reply pkg/providers/metadata/utils.go]: Fixed in b271f9a0.

**Review by jwcesign — APPROVED** [review]
